### ISSUE TITLE
Torch Alexnet Perf Investigation

### ISSUE THREAD

**Issue by gheinrich** [NVIDIA/DIGITS#687]
From @TimZaman on https://github.com/NVIDIA/DIGITS/issues/659#issuecomment-210925256:
> On Caffe it took around an hour per epoch. This seems to be taking 2 hours per epoch. But I have a 1200% luajit CPU usage now. GPU utilisation is still swinging a lot between 0 and 100 (that could be the 8tb hdd i am using).

In my experiments I found Torch Alexnet to be 30% faster than Caffe Alexnet so I'm a bit disconcerted to see that it's not the case on your system. I've only measured this on an SSD drive, where random disk accesses are not really a problem. I suppose your drive is rotational? Have you enabled shuffling? Data shuffling is ON by default for Torch but you can untick the box when creating the model and the disk read access patterns should be much more sequential. That might help.

Otherwise, do you think you'd be able to check whether we're waiting for data during training? The relevant line of code is [there](https://github.com/NVIDIA/DIGITS/blob/v3.2.0/tools/torch/main.lua#L706). You can use timers to check how long we're waiting there, e.g:
```th
local a = torch.Timer()
local accTime = 0
while (xxx)
   ...
   cutorch.synchronize()
   local t0 = a:time().real
   data_loader:waitNext()
   accTime = accTime + a:time().real - t0
   ...
end
print(string.format("waited: %.6f\n", accTime))
```

**Comment by TimZaman** [comment]
Ok i have been running tests the whole day.
I freshly started with the current digits master branch and your tds pull request alterations and your timing suggestion.
If i run on the HDD, it draws 30MB/s (100% utilization) and it forms _a bit_ of a bottleneck. I moved it onto the SSD and then it draws 60Mb/s (20% utilization) and everything is a tad faster. I calculated per batch the 'waitNext()' waiting time and it's on average around 0.001 second, as opposed to the average time per batch (Vanilla Default Digits AlexNet [batchsize 100, right?]), which is between 0.3 and 0.6 seconds. Sometimes the waiting time jumps up  once ever dozen batches and takes around 0.3 seconds.

Running AlexNet with torch i can clearly distinguish two phases; Loading and Training.
**Loading**
The loading phase takes around 5-10 minutes. The bottleneck here is probably the disk: on the SSD it draws 300Mb/s and that corresponds to the 5-10 minutes (140Gb LMDB file). It draws 0 from the GPU and it only uses a little bit of CPU-RAM, while utilizing 250% CPU.
**Training**
It draws a comfortable 60MB/s from the disk, the GPU usage is 4Gb ram and utilization swings between 0-100, temperature around 75C (would be near 94C usually when fully used). The CPU utilization however is becoming a problem here with this last `tds` inclusion. Indeed, the usage of `tds` has fixed torch from crashing now due to the 2Gb heap limitation, but it now draws the CPU fully. I really, really wonder what the CPU is so busy with?

Estimated time remaining on 1 million 256x256 images with vanilla digits alexnet takes a few days this way.

**Comment by gheinrich** [comment]
Thanks for the detailed information! Regarding the Torch measurements, was this with or without shuffling?

**Comment by TimZaman** [comment]
All without shuffling. The bottleneck does not seem to be the disk io. Also, why does it take so long (both for caffe and torch) for a job to start? I am just so puzzled of what torch is doing before the actual training. I'll dive into that more tomorrow.

**Comment by TimZaman** [comment]
Okay so this is the torch log for running vanilla digits with a masterbranch and 1 instead of 4 data threads (no tds).
This is on a 6 core, 64Gb RAM, 4x Titan X (1x is used), and proper SSD.
ImageNet 1M 256x256 images, vanilla digits standard alexnet torch.
Note the timestamps.
```
2016-04-17 20:52:56 [INFO ] Loading mean tensor from /home/tzaman/ssd_jobs/20160401-161438-510e/mean.jpg file
2016-04-17 20:52:56 [INFO ] Loading label definitions from /home/tzaman/ssd_jobs/20160401-161438-510e/labels.txt file
2016-04-17 20:52:56 [INFO ] found 998 categories
2016-04-17 20:52:56 [INFO ] creating data readers
2016-04-17 20:52:56 [INFO ] opening LMDB database: /home/tzaman/ssd_jobs/20160401-161438-510e/train_db
2016-04-17 20:56:37 [INFO ] Image channels are 3, Image width is 256 and Image height is 256
2016-04-17 20:56:37 [INFO ] found 1023461 images in train db/home/tzaman/ssd_jobs/20160401-161438-510e/train_db
2016-04-17 20:56:38 [INFO ] opening LMDB database: /home/tzaman/ssd_jobs/20160401-161438-510e/val_db
2016-04-17 20:57:56 [INFO ] Image channels are 3, Image width is 256 and Image height is 256
2016-04-17 20:57:56 [INFO ] found 255863 images in train db/home/tzaman/ssd_jobs/20160401-161438-510e/val_db
2016-04-17 20:57:57 [INFO ] Loading network definition from /home/tzaman/ssd_jobs/20160417-205254-d1d4/model
Using CuDNN backend
2016-04-17 20:57:59 [INFO ] Train batch size is 100 and validation batch size is 100
2016-04-17 20:57:59 [INFO ] Network definition:
nn.Sequential {
[input -> (1) -> (2) -> output]
(1): nn.Sequential {
[input -> (1) -> (2) -> (3) -> (4) -> (5) -> (6) -> (7) -> (8) -> (9) -> (10) -> (11) -> (12) -> (13) -> output]
(1): cudnn.SpatialConvolution(3 -> 96, 11x11, 4,4, 2,2)
(2): cudnn.ReLU
(3): cudnn.SpatialMaxPooling(3,3,2,2)
(4): cudnn.SpatialConvolution(96 -> 256, 5x5, 1,1, 2,2)
(5): cudnn.ReLU
(6): cudnn.SpatialMaxPooling(3,3,2,2)
(7): cudnn.SpatialConvolution(256 -> 384, 3x3, 1,1, 1,1)
(8): cudnn.ReLU
(9): cudnn.SpatialConvolution(384 -> 384, 3x3, 1,1, 1,1)
(10): cudnn.ReLU
(11): cudnn.SpatialConvolution(384 -> 256, 3x3, 1,1, 1,1)
(12): cudnn.ReLU
(13): cudnn.SpatialMaxPooling(3,3,2,2)
}
(2): nn.Sequential {
[input -> (1) -> (2) -> (3) -> (4) -> (5) -> (6) -> (7) -> (8) -> (9) -> output]
(1): nn.View
(2): nn.Dropout(0.500000)
(3): nn.Linear(9216 -> 4096)
(4): nn.Threshold
(5): nn.Dropout(0.500000)
(6): nn.Linear(4096 -> 4096)
(7): nn.Threshold
(8): nn.Linear(4096 -> 998)
(9): cudnn.LogSoftMax
}
}
2016-04-17 20:57:59 [INFO ] Network definition ends
2016-04-17 20:57:59 [INFO ] switching to CUDA
2016-04-17 20:58:09 [INFO ] initializing the parameters for learning rate policy: step
2016-04-17 20:58:09 [INFO ] initializing the parameters for Optimizer
2016-04-17 20:58:09 [INFO ] During training. details will be logged after every 5000 images
2016-04-17 20:58:09 [INFO ] Training epochs to be completed for each validation : 1
2016-04-17 20:58:09 [INFO ] Training epochs to be completed before taking a snapshot : 1
2016-04-17 20:58:09 [INFO ] While logging, epoch value will be rounded to 5 significant digits
2016-04-17 20:58:09 [INFO ] Model weights will be saved as snapshot_<EPOCH>_Weights.t7
2016-04-17 20:58:09 [INFO ] started training the model
2016-04-17 22:25:28 [INFO ] Validation (epoch 0): loss = 6.9204946533224, accuracy = 0.000930185294474
2016-04-17 22:25:29 [INFO ] Training (epoch 0.0001): loss = 6.9742894172668, lr = 0.01
2016-04-17 22:26:34 [INFO ] Training (epoch 0.00498): loss = 6.933155002594, lr = 0.01
2016-04-17 22:27:36 [INFO ] Training (epoch 0.00987): loss = 6.9082776260376, lr = 0.01
```
Notice the jump between 20:58:09  and 22:25:28 .. What's going on here? 1200% luajit CPU usage.

**Comment by gheinrich** [comment]
The loading time between `20:52:56` and `20:57:56` is when we're loading the LMDB keys. Unfortunately, to read LMDB keys we need to read the contents of the LMDB too. That could be sped up if we were to share keys between threads. I could do this but this sounds like a very small part of the overall train time so perhaps not something to do at a priority, right?

From `20:58:09` to `22:25:28` the framework is busy doing inference and computing accuracy on the 255863 images of your validation set. Over that time the CPU must be pretty busy decoding images (are you using an encoded LMDB - if so is this JPG or PNG?). If you have such a large validation set this is expected to take some non negligible amount of time, particularly if you have only 1 loader thread. To make this faster, you can reduce the size of the validation set or completely remove the validation set.

If you are training for the default 30 epochs then the job is expected to take a few days on Alexnet/ImageNet.

Can you summarize what you feel the issue is? Thanks!

**Comment by TimZaman** [comment]
First of all, you have definitely solved the memory issue by using tds, that's for sure, and that one can be merged i think.
My main concern now is the CPU usage at 1200%, whereas caffe uses 120% here. I rarely see high cpu usage with classifying with digits and torch.
**Using:**
DB Backend : lmdb
Image Encoding: png
DB Compression: none

**Comment by gheinrich** [comment]
Thanks for the details. The excessive CPU load is probably coming from your BLAS library. You could try doing this to limit the number of threads BLAS is going to use:
```
OMP_NUM_THREADS=1 ./digits-devserver -d
```
 Surprisingly this might not slow things down too much (it didn't in my experience).

Or you can use the `--threads` option of `main.lua` though I noticed this doesn't always work for some reason, possibly due to a build system weirdness in Torch.

**Comment by TimZaman** [comment]
So the above table suggests (i double checked the findings) that mean subtraction makes the CPU go from 350% to 1200%. Does that have anything to do with BLAS?

**Comment by gheinrich** [comment]
https://github.com/torch/torch7/blob/2243cbd0b8de8ad457b0dae1d7b8b926dcb0cb84/lib/TH/generic/THTensorMath.c#L551

**Comment by gheinrich** [comment]
Hello, do you still think there is a bug in DIGITS there?

**Comment by TimZaman** [comment]
Let me test your BLAS suggestion and threads difference tomorrow. I don't think it's a bug, but there's definitely room for improvement given the cpu usage of the mean subtraction.

**Comment by TimZaman** [comment]
Finally got around to testing it:
**opt.threads**
As you said Greg, setting opt.threads had no influence on anything. Maybe even remove it?
**OpenMP**
Setting the `OMP_NUM_THREADS` flag ofc does not influence the reader threads, so given the four loader threads and some torch overhead, the CPU usage was around 400%, as opposed to 1200%. Hence, it is indeed BLAS that's not acting nicely. It seems that this flag has no impact on speed; time/epoch.
So however this being a real issue, it can be closed as it is not digits related.
Any ideas other than using OMP_NUM_THREADS; try a different BLAS installation? which one do you use, and do you ever use the OMP_NUM_THREADS flag, Greg?

**Comment by gheinrich** [comment]
Thanks for the feedback. Torch is using OpenBLAS on my system. Without setting OMP_NUM_THREADS the CPU load is in the order of 100% times the number of reader threads so I don't set OMP_NUM_THREADS.

I'll close this issue for now, thanks.
